Tapping at the start of RTL text in bidi text selects the visual end (logical start) of the text

whsieh opened the pull request:
#### 3af28efca18605efc95068f0c3c7aaa2b010f0e6
<pre>
Tapping at the start of RTL text in bidi text selects the visual end (logical start) of the text
<a href="https://bugs.webkit.org/show_bug.cgi?id=280950">https://bugs.webkit.org/show_bug.cgi?id=280950</a>
<a href="https://rdar.apple.com/137395868">rdar://137395868</a>

Reviewed by Ryosuke Niwa.

On iOS, attempting to place the selection at the start of an RTL text run within LTR bidi text
(using any of floating text cursor, tap gesture, or clicking with a trackpad on iPadOS) causes the
selection to be placed at the logical start of the RTL text (visually, at the right edge) instead of
where you tapped/clicked.

This happens because of this iOS-specific codepath used to map coordinates to a rendered DOM
position, in `RenderText::positionForPoint`:

```
if (pointLineDirection != run-&gt;logicalLeftIgnoringInlineDirection() &amp;&amp; point.x() &lt; run-&gt;visualRectIgnoringBlockDirection().x() + run-&gt;logicalWidth()) {
    auto half = LayoutUnit { run-&gt;visualRectIgnoringBlockDirection().x() + run-&gt;logicalWidth() / 2.f };
    auto affinity = point.x() &lt; half ? Affinity::Downstream : Affinity::Upstream;
    return createVisiblePosition(offsetForPositionInRun(*run, pointLineDirection) + run-&gt;start(), affinity);
}
```

This codepath doesn&apos;t attempt to take bidi text into account at the edges of the bidi container,
causing the max and min caret offsets to be visually flipped at bidi level boundaries. To fix this,
we adopt `createVisiblePositionAfterAdjustingOffsetForBiDi` for consistency with the nearby
(platform-agnostic) return values.

* LayoutTests/editing/caret/caret-at-start-of-bidi-rtl-text-expected.txt: Added.
* LayoutTests/editing/caret/caret-at-start-of-bidi-rtl-text.html: Added.

Add a layout test to exercise the change.

* LayoutTests/platform/ios/imported/blink/editing/selection/offset-from-point-complex-scripts-expected.txt: Removed.

Remove this iOS-specific baseline, now that iOS behavior aligns with all other platforms.

* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::positionForPoint):

Canonical link: <a href="https://commits.webkit.org/284760@main">https://commits.webkit.org/284760@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c8c7c9fc3e4b3c22b893aab9d6854463a931c6c6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70409 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49815 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23174 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74498 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21587 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72526 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57615 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21427 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55794 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14262 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73475 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45315 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60695 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36256 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41974 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18131 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19948 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63905 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18475 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76217 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14635 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17712 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63515 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14677 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60761 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63451 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15591 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11488 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5122 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45617 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/384 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46691 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47968 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46433 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/rendering/RenderText.cpp:
```diff
@@ -830,8 +830,8 @@ VisiblePosition RenderText::positionForPoint(const LayoutPoint& point, HitTestSo
-                    auto affinity = point.x() < half ? Affinity::Downstream : Affinity::Upstream;
-                    return createVisiblePosition(offsetForPositionInRun(*run, pointLineDirection) + run->start(), affinity);
+                    shouldAffinityBeDownstream = point.x() < half ? AlwaysDownstream : AlwaysUpstream;
+                    return createVisiblePositionAfterAdjustingOffsetForBiDi(run, offsetForPositionInRun(*run, pointLineDirection), shouldAffinityBeDownstream);
```
Yeah... this fix comes from https://commits.webkit.org/82361@main.

whsieh committed:
Tapping at the start of RTL text in bidi text selects the visual end (logical start) of the text
https://bugs.webkit.org/show_bug.cgi?id=280950
rdar://137395868

Reviewed by Ryosuke Niwa.

On iOS, attempting to place the selection at the start of an RTL text run within LTR bidi text
(using any of floating text cursor, tap gesture, or clicking with a trackpad on iPadOS) causes the
selection to be placed at the logical start of the RTL text (visually, at the right edge) instead of
where you tapped/clicked.

This happens because of this iOS-specific codepath used to map coordinates to a rendered DOM
position, in `RenderText::positionForPoint`:

```
if (pointLineDirection != run->logicalLeftIgnoringInlineDirection() && point.x() < run->visualRectIgnoringBlockDirection().x() + run->logicalWidth()) {
    auto half = LayoutUnit { run->visualRectIgnoringBlockDirection().x() + run->logicalWidth() / 2.f };
    auto affinity = point.x() < half ? Affinity::Downstream : Affinity::Upstream;
    return createVisiblePosition(offsetForPositionInRun(*run, pointLineDirection) + run->start(), affinity);
}
```

This codepath doesn't attempt to take bidi text into account at the edges of the bidi container,
causing the max and min caret offsets to be visually flipped at bidi level boundaries. To fix this,
we adopt `createVisiblePositionAfterAdjustingOffsetForBiDi` for consistency with the nearby
(platform-agnostic) return values.

* LayoutTests/editing/caret/caret-at-start-of-bidi-rtl-text-expected.txt: Added.
* LayoutTests/editing/caret/caret-at-start-of-bidi-rtl-text.html: Added.

Add a layout test to exercise the change.

* LayoutTests/platform/ios/imported/blink/editing/selection/offset-from-point-complex-scripts-expected.txt: Removed.

Remove this iOS-specific baseline, now that iOS behavior aligns with all other platforms.

* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::positionForPoint):

Canonical link: https://commits.webkit.org/284760@main
